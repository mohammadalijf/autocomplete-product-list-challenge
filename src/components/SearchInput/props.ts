import { TextFieldProps } from "@material-ui/core";

export interface ISearchInputProps {
  /**
   * change handler for input value
   */
  onChange?: SearchInputChangeEvent;
}

/**
 * Event handler for SearchInput when input value changes
 * @param query: query string
 */
export type SearchInputChangeEvent = (query: string) => void;
/**
 * SearchInputs' props
 */
export type SearchInputProps = ISearchInputProps &
  Omit<TextFieldProps, "onChange">;
