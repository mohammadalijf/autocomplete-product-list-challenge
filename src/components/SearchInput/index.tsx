import { TextField } from "@material-ui/core";
import { ChangeEventHandler, FunctionComponent } from "react";
import { SearchInputProps, SearchInputChangeEvent } from "./props";

/**
 * Component for search input using MUI outlined TextField
 */
const SearchInput: FunctionComponent<SearchInputProps> = ({
  onChange,
  ...props
}) => {
  //#region handlers

  /**
   * ChangeEventHandler for search input
   * @param event Change event
   */
  const changeHandler: ChangeEventHandler<HTMLInputElement> = (event) => {
    onChange?.(event.target.value);
  };

  //#endregion

  return (
    <TextField
      {...props}
      autoComplete="products"
      variant="outlined"
      label="Search"
      type="search"
      inputProps={{ ...props.inputProps, "data-testid": "searchInput" }}
      onChange={changeHandler}
    />
  );
};

export default SearchInput;
export type { SearchInputProps, SearchInputChangeEvent };
