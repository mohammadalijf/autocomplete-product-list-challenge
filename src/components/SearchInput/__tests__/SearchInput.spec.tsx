import { cleanup, render, RenderResult } from "@testing-library/react";
import userEvent from "@testing-library/user-event";
import SearchInput from "..";

describe("<SearchInput />", () => {
  let searchInput: RenderResult;
  afterEach(cleanup);

  describe("typing in input", () => {
    let onChange: jest.Mock;
    beforeEach(async () => {
      onChange = jest.fn().mockName("onChange");
      searchInput = render(<SearchInput onChange={onChange} />);
    });

    it("calls the onChange handler with typed string", async () => {
      userEvent.type(searchInput.getByTestId("searchInput"), "hello");
      expect(onChange).toHaveBeenCalledWith("hello");
    });
  });
});
