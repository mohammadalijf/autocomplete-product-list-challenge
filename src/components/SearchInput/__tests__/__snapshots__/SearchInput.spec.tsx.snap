// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<SearchInput /> typing in input renders correctly 1`] = `
<DocumentFragment>
  <div
    class="MuiFormControl-root MuiTextField-root"
  >
    <label
      class="MuiFormLabel-root MuiInputLabel-root MuiInputLabel-formControl MuiInputLabel-animated MuiInputLabel-outlined"
      data-shrink="false"
    >
      Search
    </label>
    <div
      class="MuiInputBase-root MuiOutlinedInput-root MuiInputBase-formControl"
    >
      <input
        aria-invalid="false"
        autocomplete="products"
        class="MuiInputBase-input MuiOutlinedInput-input MuiInputBase-inputTypeSearch"
        data-testid="searchInput"
        type="search"
        value=""
      />
      <fieldset
        aria-hidden="true"
        class="PrivateNotchedOutline-root-1 MuiOutlinedInput-notchedOutline"
      >
        <legend
          class="PrivateNotchedOutline-legendLabelled-3"
        >
          <span>
            Search
          </span>
        </legend>
      </fieldset>
    </div>
  </div>
</DocumentFragment>
`;
